### DISCUSSION TITLE
Data race? uv__req_register vs. uv__req_unregister

### DISCUSSION THREAD

**Discussion by jan-kiszka** [libuv/libuv#3847] [Q&A] [answered]
I'm hunting ```./src/threadpool.c:329: uv__queue_done: Assertion `uv__has_active_reqs(req->loop)' failed.``` which seems to be triggered by either an imbalance/corruption of the request queue - or by a problem of the request tracking itself.

If I look at `uv__req_register` and `uv__req_unregister`, I see non-atomic inc/dec. Is that fine because both functions must always be run over the same thread? Or is some lock protecting these calls which I missed so far. If not, we have a data race here that could cause the `active_reqs.count` to become corrupt.

Relates to 88c2af0e65ed49ff599f5afd4bf952ecdfab4594 / #1746. @vtjnash

**Comment by jan-kiszka** [comment]
Indeed, this little hack seems to make my problem go away:
```
--- libuv1-1.40.0.orig/src/uv-common.h
+++ libuv1-1.40.0/src/uv-common.h
@@ -223,14 +223,14 @@ void uv__threadpool_cleanup(void);
 
 #define uv__req_register(loop, req)                                           \
   do {                                                                        \
-    (loop)->active_reqs.count++;                                              \
+    __atomic_add_fetch(&(loop)->active_reqs.count, 1, __ATOMIC_RELAXED);      \
   }                                                                           \
   while (0)
 
 #define uv__req_unregister(loop, req)                                         \
   do {                                                                        \
     assert(uv__has_active_reqs(loop));                                        \
-    (loop)->active_reqs.count--;                                              \
+    __atomic_sub_fetch(&(loop)->active_reqs.count, 1, __ATOMIC_RELAXED);      \
   }                                                                           \
   while (0)
 
```
(tested over Debian 11, therefore 1.40)

**Comment by vtjnash** [comment, marked as answer]
There is no data-race here because libuv is not permitted to be accessed from multiple threads (except explicitly allowed functions like uv_async_send)

**Reply by jan-kiszka** [reply]
Ok, now it starts to make sense. So the issue is likely in the binding lib (mraa). It uses `uv_queue_work` from a separate thread - that should be `uv_async_send` then.

**Reply by bnoordhuis** [reply]
FWIW, there's a draft PR for that (#3429) but there were some unresolved issues last I checked.
